step-api/src/main/java/org/jenkinsci/plugins/workflow/steps/AbstractSynchronousNonBlockingStepExecution.java

```diff
@@ -0,0 +1,64 @@
+package org.jenkinsci.plugins.workflow.steps;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import jenkins.util.Timer;
+
+/**
+ * {@link StepExecution} that always executes synchronously and does not block the CPS VM thread.
+ * @param <T> the type of the return value (may be {@link Void})
+ */
+public abstract class AbstractSynchronousNonBlockingStepExecution<T> extends AbstractStepExecutionImpl {
+
+    private transient volatile ScheduledFuture<?> task;
+
+    protected AbstractSynchronousNonBlockingStepExecution() {
+    }
+
+    protected AbstractSynchronousNonBlockingStepExecution(StepContext context) {
+        super(context);
+    }
+
+    /**
+     * Meat of the execution.
+     *
+     * When this method returns, a step execution is over.
+     */
+    protected abstract T run() throws Exception;
+
+    @Override
+    public final boolean start() throws Exception {
+        task = Timer.get().schedule(new StepRunner(this), 0, TimeUnit.MILLISECONDS);
```

Comment:
`Timer` is capped at ten simultaneous threads, which is good for many purposes but imposes an artificial limit here. It might be better to use a custom executor service. (Can use some or all of the various wrapper classes that `Timer` uses, as appropriate.)